File: app/src/main/java/com/mwigzell/places/dagger/AppModule.java
package com.mwigzell.places.dagger;

import android.content.Context;
import android.location.Location;

import com.mwigzell.places.model.Type;
import com.mwigzell.places.redux.AppAction;
import com.mwigzell.places.redux.AppReducer;
import com.mwigzell.places.redux.AppState;
import com.mwigzell.places.redux.ImmutableAppState;
import com.mwigzell.places.redux.PersistenceController;
import com.mwigzell.places.redux.PlaceState;
import com.mwigzell.places.redux.jedux.Logger;
import com.mwigzell.places.redux.jedux.Store.Reducer;
import com.mwigzell.places.redux.jedux.Store;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

/**
 * Created by mwigzell on 11/5/15.
 */
@Module
public class AppModule {
    private Context context;
    private Store<AppAction<Object>, AppState> store;
    private PersistenceController persistenceController;
    private AppState appState;

    public AppModule(Context context) {
        this.context = context;
        persistenceController = new PersistenceController(context);
        appState = persistenceController.getSavedState();
        if (appState == null) {
            appState = provideAppState();
        }
    }

    // need for testing, normally not for running except first time
    @Provides
    public AppState provideAppState() {
        return ImmutableAppState.builder()
                .placeState(new PlaceState())
                .state(AppState.States.INIT)
                .lastError("")
                .types(new ArrayList<Type>())
                .location(new Location("init"))
                .selectedPosition(new Integer(0))
                .build();
    }

    @Provides @Singleton
    public Store<AppAction<Object>, AppState> provideStore() {
        List<Reducer<AppAction, AppState>> reducers = new ArrayList<>();

        store =  new Store(new AppReducer(reducers), appState, new Logger("Places"), persistenceController);
        return store;
    }
}

File: app/src/main/java/com/mwigzell/places/dagger/AppModule.kt
package com.mwigzell.places.dagger

import android.content.Context
import androidx.room.Room
import com.mwigzell.places.repository.api.network.NetworkStatus
import com.mwigzell.places.repository.api.network.ServiceCreator
import com.mwigzell.places.repository.dao.*
import com.mwigzell.places.util.AndroidServices
import com.mwigzell.places.util.FileUtils
import dagger.Module
import dagger.Provides
import java.io.File
import javax.inject.Named
import javax.inject.Singleton

/**
 * Created by mwigzell on 11/5/15.
 */
@Module
class AppModule() {

    @Provides
    @Singleton
    fun provideNetworkStatus(context: Context): NetworkStatus {
        return AndroidServices(context)
    }

    @Provides
    @Named("RetrofitBaseUrl")
    fun getRetrofitBaseUrl(): String = RETROFIT_BASE_URL

    @Provides
    @Singleton
    fun providePlacesDao(placeDao: PlaceDaoGeneric): IPlaceDaoGeneric = placeDao

    @Provides
    @Named("RetrofitCacheDir")
    fun getRetrofitCacheDir(context: Context): File {
        return context.externalCacheDir!!
    }

    @Provides
    @Singleton
    fun getPlacesDatabase(context: Context): PlacesDatabase {
        return Room.databaseBuilder(context, PlacesDatabase::class.java, "PlacesDatabase").build()
    }

    @Provides
    @Singleton
    fun provideLocationDao(locationDao: LocationDaoGeneric): ILocationDaoGeneric = locationDao

    companion object {
        const val RETROFIT_BASE_URL = "https://maps.googleapis.com"
    }
}
